flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java

```diff
@@ -421,6 +509,25 @@ public void shutdown() {
 				LOG.warn("Network buffer pool did not shut down properly.", t);
 			}
 
+			for (ResultPartition[] resultPartitions : allPartitions.values()) {
+				for (ResultPartition partition : resultPartitions) {
+					partition.destroyBufferPool();
```

Comment:
Looks like, these partition/gates releases were supposed to happen on task cancel or finish.
Do you think we should have them again here? to double-check everything is released? :)

Reply:
This is a good question. 

In current implementation, if the task is finished then `NetworkEnvironment#unregisterTask` is called to release resources of partitions/gates. So this part should be retained in the case of finished task.

If the task is canceled then the partitions/gates are released directly in canceler thread and `unregisterTask` would be called finally before task exits. For `unregisterTask`, it only works for `TaskEventDispatcher` and `ResultPartitionManager`.

In future the partition/gate release should be only managed by `ShuffleService`, not exposed to outside. The canceler thread could also call `ShuffleService#unregisterTask` to release from all the components including `TaskEventDispatcher` and `ResultPartitionManager`. Then it seems reasonable to call `ShuffleService#unregisterTask` twice in task thread and canceler thread separately. What do you think or have better options?